File: counting-elements/permutation-check.rb
# Given a non-zero array of positive integers, determine if the array contains a permutation. Time complexity should be O(N).

def solution(a)

  counts, perm = Hash.new, true

  for i in 1..a.length do
    counts[i] = 0
  end

  a.each do |elem|

    if elem > a.length
      perm = false
      break
    end

    counts[elem] += 1

    if counts[elem] == 2
      perm = false
      break
    end

  end

  perm = counts.all? { |num, count| count == 1 } if !(perm)

  perm ? 1 : 0

end

File: counting-elements/permutation_check.py
# Given a non-zero array of positive integers, determine if the array contains a permutation. Time complexity should be O(N).

def solution(A):

    if len(A) == 0:
        return 0
    else:
        n = len(A) + 1
        unused_values = {}

    for i in range(1,n):
        unused_values[i] = True

    for i in A:
        if i in unused_values:
            unused_values.pop(i, None)
        else:
            return 0

    if not bool(unused_values):
        return 1
    else:
        return 0